### PR TITLE
Fix LMAXMIX default in VASP INCAR now set based on element blocks in structure

### LINKED ISSUE #3040: `LMAXMIX` should be set based on the presence of s, p, d, f electrons not based on Z
Currently, `LMAXMIX` is set as follows:

https://github.com/materialsproject/pymatgen/blob/67ba063073b468cb33b93f2dc3841fd6ae3941d5/pymatgen/io/vasp/sets.py#L529-L543

However, this is based on Z value and not based on the presence or lack of s, p, d, or f electrons (see [VASP manual](https://www.vasp.at/wiki/index.php/LMAXMIX)). For instance, Sr has Z > 20 but does not have d electrons. La has Z > 56 but does not have f electrons. And so on. We should be checking this based on the "block" of the electrons in the `Structure` object.

See https://github.com/materialsproject/atomate2/issues/364 for a mirror of this comment and the linked PR for the fix.

### PR DESCRIPTION
**Pull request by janosh**
Closes #3040.

@arosen93 Do you have a good test case for this by any chance?

### CODE REVIEW

File: pymatgen/io/vasp/sets.py [hunk 1]
```diff
@@ -526,20 +526,19 @@ def incar(self) -> Incar:
...
+        # Modify LMAXMIX if you have d or f electrons present. Note that if the user
+        # explicitly sets LMAXMIX in settings it will override this logic.
+        # Previously, this was only set if Hubbard U was enabled as per the VASP manual
+        # but following an investigation it was determined that this would lead to a
+        # significant difference between SCF -> NonSCF even without Hubbard U enabled.
+        # Thanks to Andrew Rosen for investigating and reporting.
         if "LMAXMIX" not in settings:
+            blocks = {site.specie.block for site in structure}
             # contains f-electrons
-            if any(el.Z > 56 for el in structure.composition):
```
Comment (arosen93) [inline pymatgen/io/vasp/sets.py:539]: Should be `if "f" in blocks`. I made a mistake in the WIP Atomate2 PR (fixed now).

File: pymatgen/io/vasp/sets.py [hunk 2]
```diff
@@ -526,20 +526,19 @@ def incar(self) -> Incar:
...
+            if "d" in blocks:
                 incar["LMAXMIX"] = 6
             # contains d-electrons
-            elif any(el.Z > 20 for el in structure.composition):
```
Comment (arosen93) [inline pymatgen/io/vasp/sets.py:542]: Should be `if "d" in blocks`. I made a mistake in the WIP Atomate2 PR (fixed now).

**Comment by arosen93** [comment]
I don't have a unit test for this one yet. We can copy it from the Atomate2 PR once I add it there. I have a unit test in my [Quacc code](https://github.com/Quantum-Accelerators/quacc/blob/3e2e6c2281c214ada86fcba44d43750017844a2a/tests/calculators/vasp/test_vasp.py#L63-L75), but it's not quite the same.

File: pymatgen/io/vasp/sets.py (hunk 1 above)
Comment (janosh) [reply pymatgen/io/vasp/sets.py:539]: Ah, I thought it might be swapped :)

**Comment by janosh** [comment]
@arosen93 How's this test?


```py
    def test_incar_lmaxmix(self):
        # structure containing neither f- nor d-electrons
        structure_f = self.get_structure("Si")
        assert "LMAXMIX" not in MPRelaxSet(structure_f).incar

        # structure containing d-electrons but no f-electrons
        structure_d = self.get_structure("LiFePO4")
        assert MPRelaxSet(structure_d).incar["LMAXMIX"] == 4

        # structure containing f-electrons but no d-electrons
        structure_f = structure_d.copy()
        structure_f.replace_species({"Fe": "La"})
        assert MPRelaxSet(structure_f).incar["LMAXMIX"] == 6

        # explicit LMAXMIX in settings overrides automatic selection
        structure_override = self.get_structure("Si")  # Iron is in the d-block
        set_override = MPRelaxSet(structure_override, user_incar_settings={"LMAXMIX": 3})
        assert set_override.incar["LMAXMIX"] == 3
```

**Comment by arosen93** [comment]
Looks almost right, but La doesn't actually have f electrons. Try Ce instead.

Also, it might be worth adding a test for a structure containing d and f electrons (e.g. Fe and Ce). It should set LMAXMIX = 6.

**Comment by janosh** [comment]
Are you sure?

```py
from pymatgen.core import Element

Element("La").block
>>> 'f'
```

**Comment by arosen93** [comment]
Ugh, it's an f-block element but has no f electrons... lol. https://en.wikipedia.org/wiki/Lanthanum. It's probably fine to let LMAXMIX = 6 here regardless.

**Comment by arosen93** [comment]
> Among the lanthanides, lanthanum is exceptional as it has no 4f electrons as a single gas-phase atom. Thus it is only very weakly [paramagnetic](https://en.wikipedia.org/wiki/Paramagnetic), unlike the strongly paramagnetic later lanthanides (with the exceptions of the last two, [ytterbium](https://en.wikipedia.org/wiki/Ytterbium) and [lutetium](https://en.wikipedia.org/wiki/Lutetium), where the 4f shell is completely full).[[17]](https://en.wikipedia.org/wiki/Lanthanum#cite_note-17) However, the 4f shell of lanthanum can become partially occupied in chemical environments and participate in chemical bonding

Ah okay, so testing for block is the right thing to do. Carry on with La.

**Comment by janosh** [comment]
> Also also, your final test has Si in place of Fe.

Good catch.

> Also, it might be worth adding a test for a structure containing d and f electrons (e.g. Fe and Ce). It should set LMAXMIX = 6.

Will do.

**Comment by matthewkuner** [comment]
@arosen93 @janosh would this set LMAXMIX = 4 for Gallium (a p-block element)? Because it should according to https://www.vasp.at/forum/viewtopic.php?p=24392#p24374

**Comment by arosen93** [comment]
This is an interesting point worth solidifying. In practice, I think the answer is that it depends on the pseudopotential. For instance, if one were to use `Ge`, there are no d electrons in the valence. However, if you use the (recommended) `Ge_d` then of course they are. There is never harm in "overdoing it" with regards to LMAXMIX, so I would agree that something like Ge should have LMAXMIX = 4.

What this all boils down to is that:

1. The VASP manual was ambiguous, per usual.
2. This merged PR should probably be undone to go back to the old way of checking based on Z.

**Comment by matthewkuner** [comment]
We could also just parse the electrons included in the particular pseudopotential and set LMAXMIX according to what types of electrons are present

**Comment by arosen93** [comment]
That might be smarter, albeit slightly more complex.

**Comment by matthewkuner** [comment]
This change should be reverted, no? @janosh

**Comment by janosh** [comment]
>  We could also just parse the electrons included in the particular pseudopotential and set LMAXMIX according to what types of electrons are present

@matthewkuner are you able to submit a PR for this?

**Comment by matthewkuner** [comment]
@janosh I may be able to do it eventually, but my bandwidth is a bit low rn. I think it is best to revert this change for the time being.

**Comment by arosen93** [comment]
Agreed. I think simplest option is to revert the LMAXMIX-related change. No need to overengineer a solution.

**Comment by janosh** [comment]
Just to confirm, you'd like to revert to this logic?

```py
        if "LMAXMIX" not in settings:
            # contains f-electrons
            if any(el.Z > 56 for el in structure.composition):
                incar["LMAXMIX"] = 6
            # contains d-electrons
            elif any(el.Z > 20 for el in structure.composition):
                incar["LMAXMIX"] = 4
```

**Comment by matthewkuner** [comment]
@janosh that seems good to me

**Pull request merged**